src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,250 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#include "dawn/Optimizer/PassSetBoundaryCondition.h"
+#include "dawn/Optimizer/OptimizerContext.h"
+#include "dawn/Optimizer/Stencil.h"
+#include "dawn/Optimizer/StencilInstantiation.h"
+#include "dawn/SIR/ASTExpr.h"
+#include "dawn/SIR/ASTStmt.h"
+#include "dawn/SIR/ASTUtil.h"
+#include "dawn/SIR/ASTVisitor.h"
+#include "dawn/Support/Assert.h"
+#include "dawn/Support/Logging.h"
+#include <iostream>
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+namespace dawn {
+
+/// @brief Get all field and variable accesses identifier by `AccessID`
+class GetStencilCalls : public ASTVisitorForwarding {
+  StencilInstantiation* instantiation_;
+  int StencilID_;
+
+  std::vector<std::shared_ptr<StencilCallDeclStmt>> stencilCallsToReplace_;
+
+public:
+  GetStencilCalls(StencilInstantiation* instantiation, int StencilID)
+      : instantiation_(instantiation), StencilID_(StencilID) {}
+
+  void visit(const std::shared_ptr<StencilCallDeclStmt>& stmt) override {
+    if(instantiation_->getStencilIDFromStmt(stmt) == StencilID_)
+      stencilCallsToReplace_.emplace_back(stmt);
+  }
+
+  std::vector<std::shared_ptr<StencilCallDeclStmt>>& getStencilCallsToReplace() {
+    return stencilCallsToReplace_;
+  }
+
+  void reset() { stencilCallsToReplace_.clear(); }
+};
+
+struct bcStorage {
+  std::string functor;
+  std::vector<std::string> arguments;
+};
+
+PassSetBoundaryCondition::PassSetBoundaryCondition() : Pass("PassSetBoundaryCondition") {}
+
+bool PassSetBoundaryCondition::run(StencilInstantiation* stencilInstantiation) {
+
+  // check if we need to run this pass
+  if(stencilInstantiation->getStencils().size() == 1) {
+    if(stencilInstantiation->getOptimizerContext()->getOptions().ReportBoundaryConditions) {
+      std::cout << "\nPASS: " << getName() << ": " << stencilInstantiation->getName() << " :";
+      std::cout << " No boundary conditions applied\n";
+    }
+    return true;
+  }
+  // returns the original ID of a variable since we need to make sure that if
```

Comment:
we need documentation, and I have the feeling we should move this to StencilInstantation